# Experimento 03

Replicación de la prueba realizada en el Trabajo de Diploma *Difusión de epidemias sobre redes empleando movimiento lagrangiano* para la selección del algoritmo de estimación de parámetros.

## Experiencia anterior

En dicho trabajo se emplean 3 algoritmos de optimización para la estimación de parámetros de modelos SIR con movimiento lagrangiano:
- Enjambre de Partículas
- Evolución Diferencial
- Levenberg-Marquardt

Para la realizaión de la prueba primero se resuelve un modelo con parámetros prefijados, y se extraen los resultados obtenidos. Luego se crea la función objetivo a minimizar con los mismos valores iniciales, y algunos de los parámetros anteriores utilizados, solo dejando sin emplear los parámetros que se desean buscar, haciendo esta función dependiente de estos. El autor utiliza los siguientes parámetros y valores iniciales:

- Una red de $2$ nodos conectados entre ellos
- $S_i(0) = 999$
- $I_i(0) = 1$
- $R_i(0) = 0$
- $N_i(0) = 1000$
- $\phi_{ij} = 0.5$, $\forall i \ne j$
- $\tau_{ij} = 0.5$, $\forall i \ne j$
- $\beta^* = (0.25,0.25)$
- $\gamma^* = (0.052,0.052)$

Además la resolución del SEDO se realiza para $t \in (0,300)$. En dicha prueba el autor se propone hallar los valores de $\beta$ y $\gamma$, que minimicen la función de coste, manteniendo fijos el resto de valores. Aunque no queda del todo claro, el autor en este experimento parece que asume que el valor de $\beta$ y $\gamma$ son iguales para cada nodo. La función de coste que utiliza es:

$$ f(\beta, \gamma) = \sqrt{ \frac{ \sum_{i=1}^{t} ( \widehat{I}(\beta^*, \gamma^*) - \widehat{I}(\beta, \gamma) )^2 }{t} } $$

donde $ \widehat{I}(\beta, \gamma)$ es la cantidad de infestados totales en función del tiempo, para un modelo que tiene como parámetros $\beta$ y $\gamma$, considerando además el resto de parámetros y valores iniciales prefijados anteriormente. Note entonces que $ \widehat{I}(\beta^*, \gamma^*) $ serían los infestados totales en función del tiempo, del modelo original. En otras palabras, esta función objetivo no es más que la raíz del error cuadrático medio.

El autor aprovecha además esta prueba para verificar la eficiencia de las implementaciones sin `numba` y con `numba` del evaluador del SEDO (prueba que ya se realizó en el [Experimento 2](/experiments/experiment_02/README_experiment_02.md)). Para cada algoritmo se realizan 30 simulaciones y se extrae el tiempo promedio que demora este, y la media de los errores cuadráticos medios. Los resultados obtenidos en el trabajo de diploma fueron:

| Función | tiempo promedio (seg) | error medio |
|---------|-----------------------|-------------|
| Levenberg-Marquardt | 9,517706791559855 | 0,0002232094 |
| Levenberg-Marquardt Numba | 3,115034015973409 | 0,0002232094 |
| PSO | 207,0603547811508 | 0,09694468588 |
| PSO Numba | 56,24999956289927 | 0,086242849621 |
| Differential Evolution | 569,3329263289769 | 0,00032080051 |
| Differential Evolution Numba | 103,38979012966156 | 0,00023158657 |

IMPORTANTE: estos resultados fueron logrados con una computadora con las siguientes características:
- CPU:Intel Core i5-6500, 3,20GHz
- Memoria Ram: 16 gb a 3200M Hz
- Sistema Operativo: Ubuntu 20.04,

## Replicación de la Prueba

Para repetir el experimento utilizaremos los mismos valores y la misma función objetivo, así como los mismos algoritmos, pero no aseguramos que sea exactamente la misma implementación del algoritmo de optimización.

In [237]:
from scipy.integrate import solve_ivp
from numba import njit
import matplotlib.pyplot as plt
import numpy as np
from timeit import timeit
from time import time
from statistics import mean, stdev
from sklearn.metrics import mean_squared_error

In [238]:
@njit
def fun_sir_lagrange_numba(t,y,Out, In, Beta, Gamma):
    
    K = Out.shape[0]
    Out_i_k = Out.sum(axis=1)

    y = y.reshape((4,K,K))
    I_k_i = y[1].sum(axis=0)
    N_k_i = y[3].sum(axis=0)
    new_y = np.zeros((4,K,K))
    for i in range(K):
        for j in range(K):
            if i == j:
                new_y[0,i,i] = - Beta[i] * y[0,i,i] * I_k_i[i] / N_k_i[i] - \
                                y[0,i,i] * Out_i_k[i] + (In[i,] * y[0,i]).sum()
                new_y[1,i,i] =   Beta[i] * y[0,i,i] * I_k_i[i] / N_k_i[i] - Gamma[i] * y[1,i,i] - \
                                y[1,i,i] * Out_i_k[i] + (In[i,] * y[1,i]).sum()
                new_y[2,i,i] =   Gamma[i] * y[1,i,i] - \
                                y[2,i,i] * Out_i_k[i] + (In[i,] * y[2,i]).sum()
                new_y[3,i,i] = - y[3,i,i] * Out_i_k[i] + (In[i,] * y[3,i]).sum()
            else:
                new_y[0,i,j] = - Beta[j] * y[0,i,j] * I_k_i[j] / N_k_i[j] - \
                                In[i,j] * y[0,i,j] + Out[i,j] * y[0,i,i]
                new_y[1,i,j] =   Beta[j] * y[0,i,j] * I_k_i[j] / N_k_i[j] - Gamma[j] * y[1,i,j] - \
                                In[i,j] * y[1,i,j] + Out[i,j] * y[1,i,i]
                new_y[2,i,j] =   Gamma[j] * y[1,i,j] - \
                                In[i,j] * y[2,i,j] + Out[i,j] * y[2,i,i]
                new_y[3,i,j] = - In[i,j] * y[3,i,j] + Out[i,j] * y[3,i,i]

    new_y = new_y.reshape((4*K*K,))
    return new_y

In [239]:
Beta = np.array([0.25, 0.25])
Gamma = np.array([0.052,0.052])
Out = np.array([
    [0, 0.5],
    [0.5,0]
])
In = Out.copy()
y0 = np.zeros((4,2,2))
np.fill_diagonal(y0[0], 999)
np.fill_diagonal(y0[1], 1)
np.fill_diagonal(y0[3], 1000)
y0 = y0.flatten()
ts = np.linspace(0,300,301)

real_sol = solve_ivp(fun_sir_lagrange_numba,(0,300), y0, t_eval=ts, args=(Out, In, Beta, Gamma))
real_y = real_sol.y.reshape((4, 2*2, real_sol.y.shape[1]))
real_I = real_y[1].sum(axis=0)

In [291]:
def gen_individual_table(values):
    text = f'| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |'
    text += f'\n|---|---|---|---|---|---|\n| '
    means = np.round(values.mean(axis=0), decimals=6)
    stds = np.round(values.std(axis=0), decimals=6)
    for i in range(3):
        text += str(means[i]) + ' | '
        text += str(stds[i]) + ' | '
    return text

### Enjambre de Partículas

El algoritmo de enjambre de partículas (PSO, por sus siglas en inglés) es una metaheurística inspirada en el comportamiento social de los pájaros o peces en un enjambre. El algoritmo PSO se basa en la idea de que las partículas en un enjambre pueden colaborar para buscar soluciones de alta calidad en el espacio de búsqueda. Para entender este algoritmo primero se debe tener en cuenta sus características y consideraciones:

- **Partículas**: Cada solución candidata a un problema de optimización se representa como una partícula en el espacio de búsqueda. Cada partícula tiene una posición y una velocidad.

- **Posición y Velocidad**: La posición de una partícula representa una solución en el espacio de búsqueda, y la velocidad de la partícula determina cómo se mueve en ese espacio. La posición y la velocidad de las partículas se actualizan en cada iteración del algoritmo.

- **Mejor Posición Personal (pbest)**: Cada partícula lleva un registro de su mejor posición individual (pbest), que es la mejor solución que ha encontrado hasta el momento en su recorrido.

- **Mejor Posición Global (gbest)**: En el enjambre, existe una partícula que ha encontrado la mejor solución entre todas las partículas. Esta es la mejor posición global (gbest) y es compartida entre todas las partículas.

- **Movimiento de las Partículas**: El movimiento de cada partícula se rige por tres influencias principales:

    - **Inercia**: Cada partícula mantiene una inercia que determina cuánta importancia se da a la velocidad actual. Este indicador suele ser representado por $w$.
    - **Atracción hacia pbest**: Las partículas son atraídas hacia su mejor posición personal (pbest). Este indicador se denota por $c_1$.
    - **Atracción hacia pbest**: Similar al anterior pero esta vez, son atraídas hacia la mejor posición global. Se denota por $c_2$

- **Actualización de las Posiciones y Velocidades**: En cada iteración, se actualizan las posiciones y velocidades de todas las partículas siguiendo las reglas de movimiento basadas en la inercia y las atracciones hacia pbest y gbest.

- **Terminación**: El algoritmo continúa iterando hasta que se cumple un criterio de terminación, que podría ser un número máximo de iteraciones, una convergencia satisfactoria o un valor de la función objetivo deseado.

Explicado este restaría analizar el funcionamiento de las ecuaciones de actualización de Posición y de Velocidad. Para ellos denotemos por $i$ a la $i-$ésima partícula del enjambre, $v_i$ su velocidad, $x_i$ su posición, $pbest_i$ la mejor posición de la partícula $i$, $gbest$ la mejor posición global alcanzada en el enjambre, $U_1, U_2 \sim U(0,1)$ dos vectores con ditribución Uniforme, $w$ coeficiente de Inercia, $c_1$ coeficiente de atracción personal, y $c_2$ coeficiente de atracción global.

Luego la ecuación para actualizar la velocidad de una partícula $i$ es:

$$ v_i^{NEW} = w v_i + c_1 U_1 \otimes (pbest_i - x_i) + c_2 U_2 \otimes ( gbest - x_i ) $$

y la ecuación para actualizar la nueva posición sería:

$$ x_i^{NEW} = x_i + v_i^{NEW} $$

La primera versión de este algoritmo no incluía el coeficiente de inercia, lo que sería equivalente a considerar $w=1$. Entendiendo lo descrito anteriormente se pueden inferir lo siguiente:

- Un aumento de la inercia $w$, favorece la exploración, mientras que una disminución de este favorece la explotación.
- Favorecer más la atracción personal $c_1$ incita a la exploración, en cambio favorecer la atracción global incrementará la explotación.

Globalmente se suelen usar los valores de estos parámetros, donde se favorecen de igualforma la atracción personal y global, equilibrando explotación y exploración, sin embargo se selecciona un valor cercano a $1$ pero por debajo de este, para con el paso del tiempo favorecer la convergencia hacia la mejor posición global:
- $w = 0.8$
- $c_1 = 0.5$
- $c_2 = 0.5$

Formalizando el este algoritmo, el seudocódigo sería el siguiente:

```
REPEAT
    FOR each particle i
        IF f (xi ) > f (pbest_i ) THEN pbest_i ← xi ;
            gbest ← max(pneighbors );
            Update velocity (NEW_v_i);
            Update position (NEW_x_i);
    END
UNTIL termination criterion is met;
```

#### Probando PSO Versión 1

En este caso usaremos el módulo `sko.PSO` que nos ofrece la clase `PSO` que implementa este algoritmo. Usaremos los siguientes parámetros:

- Cantidad de partículas: 15
- Cantidad de iteraciones máximas: 100
- $w = 0.8$
- $c_1 = 0.5$
- $c_2 = 0.5$
- Condición de Parada: Se añade que cuando `fitness(x) < 0.1` se detenga el algoritmo

In [282]:
from sko.PSO import PSO

def fitness_PSO(value):
    new_Beta = np.array([value[0], value[0]])
    new_Gamma = np.array([value[1], value[1]])
    estimate_sol = solve_ivp(fun_sir_lagrange_numba,(0,300), y0, t_eval=ts, args=(Out, In, new_Beta, new_Gamma))
    estimate_y = estimate_sol.y.reshape((4, 2*2, estimate_sol.y.shape[1]))
    estimate_I = estimate_y[1].sum(axis=0)
    return mean_squared_error(real_I, estimate_I)

In [289]:
'''
results_PSO_1 = np.zeros((30,3))

def run_pso_1(i):
    pso = PSO(fitness_PSO, 2, 15, 100, 0.0, 1.0, verbose=False)
    pso.run(1)
    iter = 1
    while pso.best_y[0] > 0.1 and iter < 100:
        iter +=1
        pso.run(1)
    results_PSO_1
    results_PSO_1[i,1] = pso.best_y
    results_PSO_1[i,2] = iter

for i in range(30):
    results_PSO_1[i,0] = timeit(lambda: run_pso_1(i), number=1)
np.savez_compressed('results_PSO_1', results_PSO_1=results_PSO_1)
'''

loaded_PSO_1 = np.load('results_PSO_1.npz', allow_pickle=True)
results_PSO_1 = loaded_PSO_1['results_PSO_1']

In [292]:
print(gen_individual_table(results_PSO_1))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 16.146225 | 2.786821 | 0.042763 | 0.024856 | 47.766667 | 7.636244 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 16.146225 | 2.786821 | 0.042763 | 0.024856 | 47.766667 | 7.636244 | 

#### Probando PSO Versión 2

En este caso usaremos el módulo `sko.PSO` que nos ofrece la clase `PSO` que implementa este algoritmo. Usaremos los siguientes parámetros:

- Cantidad de partículas: 15
- Cantidad de iteraciones máximas: 100
- $w = 0.8$
- $c_1 = 0.7$
- $c_2 = 0.3$
- Condición de Parada: Se añade que cuando `fitness(x) < 0.1` se detenga el algoritmo

Note que estamos favoreciendo la atracción personal (exploración) por encima de la atracción global (explotación).

In [255]:
'''
results_PSO_2 = np.zeros((30,3))

def run_pso_2(i):
    pso = PSO(fitness_PSO, 2, 15, 100, 0.0, 1.0, verbose=False,c1=0.7, c2=0.3)
    pso.run(1)
    iter = 1
    while pso.best_y[0] > 0.1 and iter < 100:
        iter +=1
        pso.run(1)
    results_PSO_2
    results_PSO_2[i,1] = pso.best_y
    results_PSO_2[i,2] = iter

for i in range(30):
    results_PSO_2[i,0] = timeit(lambda: run_pso_2(i), number=1)
np.savez_compressed('results_PSO_2', results_PSO_2=results_PSO_2)
'''

loaded_PSO_2 = np.load('results_PSO_2.npz', allow_pickle=True)
results_PSO_2 = loaded_PSO_2['results_PSO_2']

In [293]:
print(gen_individual_table(results_PSO_2))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 17.973089 | 2.223733 | 0.047235 | 0.032874 | 51.8 | 5.81607 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 17.973089 | 2.223733 | 0.047235 | 0.032874 | 51.8 | 5.81607 | 

#### Probando PSO Versión 3

En este caso usaremos el módulo `sko.PSO` que nos ofrece la clase `PSO` que implementa este algoritmo. Usaremos los siguientes parámetros:

- Cantidad de partículas: 15
- Cantidad de iteraciones máximas: 100
- $w = 0.8$
- $c_1 = 0.3$
- $c_2 = 0.7$
- Condición de Parada: Se añade que cuando `fitness(x) < 0.1` se detenga el algoritmo

Note que estamos favoreciendo la atracción global (explotación) por encima de la atracción personal(exploración).

In [295]:
#'''
results_PSO_3 = np.zeros((30,3))

def run_pso_3(i):
    pso = PSO(fitness_PSO, 2, 15, 100, 0.0, 1.0, verbose=False,c1=0.3, c2=0.7)
    pso.run(1)
    iter = 1.0
    while pso.best_y[0] > 0.1 and iter < 100:
        iter +=1.0
        pso.run(1)
    results_PSO_3
    results_PSO_3[i,1] = pso.best_y
    results_PSO_3[i,2] = iter

for i in range(30):
    results_PSO_3[i,0] = timeit(lambda: run_pso_3(i), number=1)
    print(i)
np.savez_compressed('results_PSO_3', results_PSO_3=results_PSO_3)
#'''

loaded_PSO_3 = np.load('results_PSO_3.npz', allow_pickle=True)
results_PSO_3 = loaded_PSO_3['results_PSO_3']

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29


In [296]:
print(gen_individual_table(results_PSO_3))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 20.056115 | 8.733108 | 4898.532286 | 18328.467931 | 55.866667 | 12.940204 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 32.002476 | 1.023986 | 1e-05 | 1.8e-05 | 100.0 | 0.0 | 

### Evolución Diferencial

<span style="color:red">FALTA agregar información sobre este algoritmo!!!</span>

In [243]:
from sko.DE import DE

def fitness_DE(value):
    new_Beta = np.array([value[0], value[0]])
    new_Gamma = np.array([value[1], value[1]])
    estimate_sol = solve_ivp(fun_sir_lagrange_numba,(0,300), y0, t_eval=ts, args=(Out, In, new_Beta, new_Gamma))
    estimate_y = estimate_sol.y.reshape((4, 2*2, estimate_sol.y.shape[1]))
    estimate_I = estimate_y[1].sum(axis=0)
    return mean_squared_error(real_I, estimate_I)

In [297]:
'''
results_DE_1 = np.zeros((30,3))

def run_de_1(i):
    de = DE(fitness_DE, n_dim=2, size_pop=14, max_iter=100, lb=[0.0,0.0], ub=[1.0,1.0])
    de.run(1)
    iter = 1
    while de.best_y[0] > 0.1 and iter < 100:
        iter +=1.0
        de.run(1)
    results_DE_1
    results_DE_1[i,1] = de.best_y
    results_DE_1[i,2] = iter

for i in range(30):
    results_DE_1[i,0] = timeit(lambda: run_de_1(i), number=1)
np.savez_compressed('results_DE_1', results_DE_1=results_DE_1)
'''

loaded_DE_1 = np.load('results_DE_1.npz', allow_pickle=True)
results_DE_1 = loaded_DE_1['results_DE_1']

In [298]:
print(gen_individual_table(results_DE_1))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 57.05207 | 8.91154 | 1.773334 | 7.635237 | 88.733333 | 12.688402 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 57.05207 | 8.91154 | 1.773334 | 7.635237 | 88.733333 | 12.688402 |

In [299]:
'''
results_DE_2 = np.zeros((30,3))

def run_de_2(i):
    de = DE(fitness_DE, n_dim=2, size_pop=14, max_iter=100, lb=[0.0,0.0], ub=[1.0,1.0],prob_mut=0.7,F=0.5)
    de.run(1)
    iter = 1
    while de.best_y[0] > 0.1 and iter < 100:
        iter +=1.0
        de.run(1)
    results_DE_2
    results_DE_2[i,1] = de.best_y
    results_DE_2[i,2] = iter

for i in range(30):
    results_DE_2[i,0] = timeit(lambda: run_de_2(i), number=1)
np.savez_compressed('results_DE_2', results_DE_2=results_DE_2)
'''

loaded_DE_2 = np.load('results_DE_2.npz', allow_pickle=True)
results_DE_2 = loaded_DE_2['results_DE_2']

In [300]:
print(gen_individual_table(results_DE_2))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 23.522002 | 7.11211 | 98.931295 | 532.521319 | 38.7 | 12.122019 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 23.522002 | 7.11211 | 98.931295 | 532.521319 | 38.7 | 12.122019 | 

In [248]:
from scipy.optimize import differential_evolution
'''
results_DE_3 = np.zeros((30,3))

def run_de_3(i):
    result = differential_evolution(fitness_DE, [(0, 1), (0, 1)],
                                maxiter=100,disp=False,polish=False,
                                workers=1, callback=lambda xk, convergence: fitness_DE(xk) < 0.1)
    results_DE_3[i,1] = result.fun
    results_DE_3[i,2] = result.nit

for i in range(30):
    results_DE_3[i,0] = timeit(lambda: run_de_3(i), number=1)
np.savez_compressed('results_DE_3', results_DE_3=results_DE_3)
'''

loaded_DE_3 = np.load('results_DE_3.npz', allow_pickle=True)
results_DE_3 = loaded_DE_3['results_DE_3']

In [301]:
print(gen_individual_table(results_DE_3))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 13.216901 | 1.746068 | 0.049754 | 0.031915 | 19.5 | 2.741654 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 13.216901 | 1.746068 | 0.049754 | 0.031915 | 19.5 | 2.741654 | 

In [250]:
'''
results_DE_4 = np.zeros((30,3))

def run_de_4(i):
    result = differential_evolution(fitness_DE, [(0, 1), (0, 1)],
                                maxiter=100,disp=False,polish=False, updating='deferred',
                                workers=-1, callback=lambda xk, convergence: fitness_DE(xk) < 0.1)
    results_DE_4[i,1] = result.fun
    results_DE_4[i,2] = result.nit

for i in range(30):
    results_DE_4[i,0] = timeit(lambda: run_de_4(i), number=1)
np.savez_compressed('results_DE_4', results_DE_4=results_DE_4)
'''

loaded_DE_4 = np.load('results_DE_4.npz', allow_pickle=True)
results_DE_4 = loaded_DE_4['results_DE_4']

In [302]:
print(gen_individual_table(results_DE_4))

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 7.944636 | 0.861948 | 0.047819 | 0.030119 | 22.9 | 2.612151 | 


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 7.944636 | 0.861948 | 0.047819 | 0.030119 | 22.9 | 2.612151 | 

In [385]:
from scipy.optimize import least_squares
from scipy.special import expit, logit

def fitness_LS(value):
    new_Beta = expit([value[0], value[0]])
    new_Gamma = expit([value[1], value[1]])
    estimate_sol = solve_ivp(fun_sir_lagrange_numba,(0,300), y0, t_eval=ts, args=(Out, In, new_Beta, new_Gamma))
    estimate_y = estimate_sol.y.reshape((4, 2*2, estimate_sol.y.shape[1]))
    estimate_I = estimate_y[1].sum(axis=0)
    return real_I - estimate_I

In [410]:
'''
results_LM_1 = np.zeros((30,4))

def run_lm_1(i):
    result = least_squares(fitness_LS, logit(np.random.random((2,))), method='lm')
    results_LM_1[i,1] = result.cost
    results_LM_1[i,2] = 1
    results_LM_1[i,3] = result.nfev

for i in range(30):
    results_LM_1[i,0] = timeit(lambda: run_lm_1(i), number=1)

np.savez_compressed('results_LM_1', results_LM_1=results_LM_1)
'''

loaded_LM_1 = np.load('results_LM_1.npz', allow_pickle=True)
results_LM_1 = loaded_LM_1['results_LM_1']

In [411]:
print(gen_individual_table(results_LM_1))
print(results_LM_1.mean(axis=0)[3])
print(results_LM_1.std(axis=0)[3])

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 1.184774 | 1.758653 | 4119849.723321 | 5414559.389893 | 1.0 | 0.0 | 
38.56666666666667
54.27503006806066


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion | Cant. Evaluaciones Promedio | Cant. Evaluaciones Desviacion |
|---|---|---|---|---|---|---|---|
| 1.184774 | 1.758653 | 4119849.723321 | 5414559.389893 | 1.0 | 0.0 | 38.566667 | 54.275030 |

In [413]:
'''
results_LM_2 = np.zeros((30,4))

def run_lm_2(i):
    result = least_squares(fitness_LS, logit([0.5,0.5]), method='lm')
    results_LM_2[i,1] = result.cost
    results_LM_2[i,2] = 1
    results_LM_2[i,3] = result.nfev

for i in range(30):
    results_LM_2[i,0] = timeit(lambda: run_lm_2(i), number=1)

np.savez_compressed('results_LM_2', results_LM_2=results_LM_2)
'''

loaded_LM_2 = np.load('results_LM_2.npz', allow_pickle=True)
results_LM_2 = loaded_LM_2['results_LM_2']

In [414]:
print(gen_individual_table(results_LM_2))
print(results_LM_2.mean(axis=0)[3])
print(results_LM_2.std(axis=0)[3])

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 1.164852 | 0.103465 | 0.0 | 0.0 | 1.0 | 0.0 | 
40.0
0.0


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion | Cant. Evaluaciones Promedio | Cant. Evaluaciones Desviacion |
|---|---|---|---|---|---|---|---|
| 1.164852 | 0.103465 | 0.0 | 0.0 | 1.0 | 0.0 | 40.0 | 0.0 |

In [417]:
'''
results_LM_3 = np.zeros((30,4))

def run_lm_3(i):
    for _ in range(10):
        result = least_squares(fitness_LS, logit(np.random.random((2,))), method='lm')
        results_LM_3[i,2] += 1
        results_LM_3[i,3] += result.nfev
        if result.cost < 0.1:
            results_LM_3[i,1] = result.cost
            return
    results_LM_3[i,1] = result.cost

for i in range(30):
    results_LM_3[i,0] = timeit(lambda: run_lm_3(i), number=1)

np.savez_compressed('results_LM_3', results_LM_3=results_LM_3)
'''

loaded_LM_3 = np.load('results_LM_3.npz', allow_pickle=True)
results_LM_3 = loaded_LM_3['results_LM_3']

In [418]:
print(gen_individual_table(results_LM_3))
print(results_LM_3.mean(axis=0)[3])
print(results_LM_3.std(axis=0)[3])

| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|
| 1.180474 | 0.813527 | 0.0 | 0.0 | 1.566667 | 1.202313 | 
41.86666666666667
28.90297946963638


| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion | Cant. Evaluaciones Promedio | Cant. Evaluaciones Desviacion |
|---|---|---|---|---|---|---|---|
| 1.180474 | 0.813527 | 0.0 | 0.0 | 1.566667 | 1.202313 | 41.866667 | 28.902979 |

In [419]:
def gen_total_table(array_values, names):
    text = f'| Métodos| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |'
    text += f'\n|---|---|---|---|---|---|--|--|\n| '
    for i, values in enumerate(array_values):
        text += f'| {names[i]} | '
        means = np.round(values.mean(axis=0), decimals=6)
        stds = np.round(values.std(axis=0), decimals=6)
        for j in range(3):
            text += str(means[j]) + ' | '
            text += str(stds[j]) + ' | '
        text += ' - | - |\n'
    return text

In [422]:
results = [
    results_PSO_1,
    results_PSO_2,
    results_PSO_3,
    results_DE_1,
    results_DE_2,
    results_DE_3,
    results_DE_4,
    results_LM_1,
    results_LM_2,
    results_LM_3
]
names = [
    'PSO 1',
    'PSO 2',
    'PSO 3',
    'DE 1',
    'DE 2',
    'DE 3',
    'DE 4',
    'LM 1',
    'LM 2',
    'LM 3'
]
print(gen_total_table(results, names))

| Métodos| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion |
|---|---|---|---|---|---|--|--|
| | PSO 1 | 16.146225 | 2.786821 | 0.042763 | 0.024856 | 47.766667 | 7.636244 |  - | - |
| PSO 2 | 17.973089 | 2.223733 | 0.047235 | 0.032874 | 51.8 | 5.81607 |  - | - |
| PSO 3 | 20.056115 | 8.733108 | 4898.532286 | 18328.467931 | 55.866667 | 12.940204 |  - | - |
| DE 1 | 57.05207 | 8.91154 | 1.773334 | 7.635237 | 88.733333 | 12.688402 |  - | - |
| DE 2 | 23.522002 | 7.11211 | 98.931295 | 532.521319 | 38.7 | 12.122019 |  - | - |
| DE 3 | 13.216901 | 1.746068 | 0.049754 | 0.031915 | 19.5 | 2.741654 |  - | - |
| DE 4 | 7.944636 | 0.861948 | 0.047819 | 0.030119 | 22.9 | 2.612151 |  - | - |
| LM 1 | 1.184774 | 1.758653 | 4119849.723321 | 5414559.389893 | 1.0 | 0.0 |  - | - |
| LM 2 | 1.164852 | 0.103465 | 0.0 | 0.0 | 1.0 | 0.0 |  - | - |
| LM 3 | 1.180474 | 0.813527 | 0.0 | 0.0 | 1.566667 | 1.202313 |  - | - |



| Métodos| Tiempo Promedio (seg) | Tiempo Desviacion (seg) | RMSE Promedio | RMSE Desviacion | Iteraciones Promedio | Iteraciones Desviacion | Cant. Evaluaciones Promedio | Cant. Evaluaciones Desviacion |
|---|---|---|---|---|---|---|---|---|
| PSO 1 | 16.146225 | 2.786821 | 0.042763 | 0.024856 | 47.766667 | 7.636244 |  - | - |
| PSO 2 | 17.973089 | 2.223733 | 0.047235 | 0.032874 | 51.8 | 5.81607 |  - | - |
| PSO 3 | 20.056115 | 8.733108 | 4898.532286 | 18328.467931 | 55.866667 | 12.940204 |  - | - |
| DE 1 | 57.05207 | 8.91154 | 1.773334 | 7.635237 | 88.733333 | 12.688402 |  - | - |
| DE 2 | 23.522002 | 7.11211 | 98.931295 | 532.521319 | 38.7 | 12.122019 |  - | - |
| DE 3 | 13.216901 | 1.746068 | 0.049754 | 0.031915 | 19.5 | 2.741654 |  - | - |
| DE 4 | 7.944636 | 0.861948 | 0.047819 | 0.030119 | 22.9 | 2.612151 |  - | - |
| LM 1 | 1.184774 | 1.758653 | 4119849.723321 | 5414559.389893 | 1.0 | 0.0 | 38.566667 | 54.275030 |
| LM 2 | 1.164852 | 0.103465 | 0.0 | 0.0 | 1.0 | 0.0 | 40.0 | 0.0 |
| LM 3 | 1.180474 | 0.813527 | 0.0 | 0.0 | 1.566667 | 1.202313 | 41.866667 | 28.902979 |